src/psyclone/domain/lfric/kern_call_arg_list.py

```diff
@@ -279,6 +303,27 @@ def stencil(self, arg, var_accesses=None):
                                      self._cell_ref_name(var_accesses))
         self.append(name, var_accesses, var_access_name=var_name)
 
+    def stencil_2d(self, arg, var_accesses=None):
+        '''Add general 2D stencil information associated with the argument 'arg'
+        to the argument list. If supplied it also stores this access in
+        var_accesses.
+
+        :param arg: the meta-data description of the kernel \
+            argument with which the stencil is associated.
+        :type arg: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
+        :param var_accesses: optional VariablesAccessInfo instance to store \
+            the information about variable accesses.
+        :type var_accesses: \
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        '''
+        # add in stencil dofmap
+        from psyclone.dynamo0p3 import DynStencils
+        var_name = DynStencils.dofmap_name(self._kern.root.symbol_table, arg)
+        name = "{0}(:,:,:,{1})".format(var_name,
```

Comment:
Mainly to help my understanding: please could you explain why we're passing a 3D array for a 2D stencil (and similarly, I notice we're already passing a 2D array for a 1D stencil), even allowing for the fact that we need to index in to the stencil for the current cell. Either I'm missing something obvious or our documentation is a bit lacking.

Reply:
The array size passed for the stencil dofmaps is laid out like so: stencil_dofmap(dof in cell, cell in stencil, main cell) where main cell is obviously the cell the kernel is being called from the PSylayer, this is how the older stencils (e.g. cross, 1dx, 1dy, etc.) are passed. The problems with this (and the reason for the creation of the 2D cross stencil) is that when used on a planar mesh instead of a cubedsphere - as is the case with the limited area meshes - the size of the stencil at the boundaries changes, this is fine as long as you don't care which direction the stencil cell is in relation to the centre cell. If you do care about the direction (as is the case when calculating winds) you can't guarentee that a cell in the stencil is in a particular direction. To overcome this the 2D cross stencil bakes that information into the stencil dofmap, so: stencil_dofmap(dof in cell, cell in branch, branch in stencil, main cell) thus we have the extra dimension for the branch of the stencil which are always in the same order W,S,E,N. There are some pictures that might give you a better idea here: https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicInfrastructure/StencilDOFMapDesign  

Hopefully this is helpful